core/configuration.py

```diff
@@ -12,13 +12,19 @@ class ConfigurationManager(object):
 	__instance = None
 
 	@staticmethod
-	def getInstance():
-		""" Static access method. """
-		if ConfigurationManager.__instance == None:
-			ConfigurationManager()
+	def get_instance(config_filepath="core/configuration.ini", input_function=None):
```

Comment:
So the design change of having these two parameters on the `get_instance()` function introduces a couple problems:

1. Whenever we use a `config_filepath` other than the default, each consumer of the Configuration Manager needs to know, store, and pass around this variable (which makes it cumbersome and exposes test-related variables and logic across the entire codebase).

2. Bootstrap is called **not only** once during initialization, but every time during the retrieval of the singleton (which is inefficient).

To solve these, I propose we make the `bootstrap()` function take in the `config_filepath` and `input_function`, the `get_instance()` function take no parameters, and that we make a getter for the `config` property that raises an exception if the class hasn't been bootstrapped yet.

At the end of the day we just need to bootstrap the Configuration Manager only once during the very beginning of the service, so this works in accordance to that. For tests, just bootstrap right after you call `get_instance()` for the first time (or every time you call the instance getter if you're creating new CM's for every single test).

Reply:
_Summarizing what I said in meeting for clarity_

I revamped the design of the class. `__init__` takes optional parameters for `config_filepath` and `input_function`, which are attributes of the instance. `bootstrap` and `get_instance` take no parameters. So from a consumer's point of view (as long as the file path does not change), all that needs to be done is initialize the class and call `get_instance` to retrieve the instance.

The tests are a bit more complex. So in order to uphold all the design suggestions you made (singleton, keep `bootstrap` in `__init__`, don't use test variables, etc.) I decided to use fixtures. Each test file creates/alters a `ConfigurationManager` instance so that it knows where to pull/create the config file. The only exception is `DMLScheduler`, where all of that happens outside of the test methods and fixtures because the runners never finish running otherwise. Not sure if this worth worrying about for now, but I can create a TODO card to try to debug why fixtures don't work for the scheduler at a later point. 

Tl;dr tests pass and from a consumer standpoint, I believe I've addressed all your concerns.